syntax = "proto3";

service Strelka {
  rpc StreamData(stream ScanData) returns (ScanResult) {}

  rpc SendLocation(ScanLocation) returns (ScanResult) {}
}

// contains request details related to the RPC
message Request {
  string uid = 1;
  string client = 2;
  string source = 3;
}

// contains metadata related to the Data or Location
message Metadata {
  string filename = 1;
  repeated string flavors = 2;
  map<string, string> metadata = 3;
}

// contains parameters that enable event retrieval for the client
message Retrieve {
  bool bundle = 1;
  string case = 2;
}

message ScanData {
  bytes data = 1; // required
  Metadata metadata = 2;
  Request request = 3;
  Retrieve retrieve = 4;
}

message ScanLocation {
  map<string, string> location = 1;  // required
  Metadata metadata = 2;
  Request request = 3;
  Retrieve retrieve = 4;
}

message ScanResult {
  string uid = 1;
  repeated string events = 2;
}
